### DISCUSSION TITLE
Tooltip on Trigger

### DISCUSSION THREAD

**Discussion by PhilGarb** [radix-ui/primitives#560] [Help]
Hi everyone,

I have a problem when using the Tooltip together with the Collapsible Trigger. I am styling the trigger based on data-state="open" to signal the open state to the user. When wrapping the Trigger in a custom Tooltip component, as described here: [Radix Docs](https://radix-ui.com/primitives/docs/components/tooltip#custom-apis), the data-state attribute is overwritten by the Tooltips data-state.

I think a Tooltip should not really interfere in this way. I have tried to remove the Slot, but then I have two buttons both with their own data-state. This does however interfere with styling since in this case I have to change my styling to accommodate the deeper nesting. 

I think the root cause is that all state is represented by only one attribute data-state and no way to distinguish states of different elements. Is there any way to make this work? Please let me know if you need a reproduction.

**Comment by benoitgrelard** [comment]
You are 100% correct, this is a clash because they both use `data-state` attribute. Few people have encountered this so far as it's an edge case when combining 2 primitives which happen to have state. This is a tricky problem to solve correctly.

The easy solution is to namespace the state attributes but that would be a huge restriction on DX for all of the most common use cases.

We'll need to think about this a lot more I think. In the meantime if you have some ideas/suggestions feel free to let us know!

**Reply by PhilGarb** [reply]
I agree it is tricky and might not be worth it to complicate the API (and make changes to basically all state-full radix components) for this. I think the main issue here is, that radix contains quite a few components that dynamically hide content, like Collapsibles, Tabs or Dialogs (basically all the components that have a trigger). In my opinion these benefit from proper Tooltip support, because they are often only represented by an icon or button with a very short label. 

Maybe the **Slot component** is to blame here? The Slot takes a component and replaces it. Shouldn't it preserve the state of the component it slots in? I am wondering, because maybe this issue is actually just a problem when replacing and we could make the Slot component namespace the state attributes. This way the general API remains the same and only the process of actually combining the two elements would be a little more complex (although I would argue, that it feels quite natural that the Slot component "merges" the props instead of replacing them and therefore the state becomes namespaced).

**Reply by PhilGarb** [reply]
Maybe the merging of the props in the Slot component could more intelligently handle the case of two data-state attributes.

https://github.com/radix-ui/primitives/blob/cf4dae46be258fc420050f58eca8efca574cd862/packages/react/slot/src/Slot.tsx#L74-L90

It could check for the presence of two data attributes and then namespace each one with the components name.

**Reply by benoitgrelard** [reply]
I don't think it is, the issue would be the same if you compose things this way:

```tsx
<Tooltip.Trigger as={Collapsible.Button}>
```

The issue is that both of these have a `data-state`, so although we always merge everything, inevitably one will override the other here.

**Reply by PhilGarb** [reply]
Yeah you are right. That was just an idea 😅. I think for now I am just not relying on the data attribute for styling. I have solved my problem by controlling the component myself and apply styling based on the local state variable. That way I can have all the differentiation between states I need.

**Reply by benoitgrelard** [reply]
That's good to know. That is something we will probably have to think about more, as we add more and more components.

**Reply by PhilGarb** [reply]
I have maybe found another option. There is the option to match only part of the data attributes when they are space separated. So something like: `data-state="tooltip-open collapsibile-open"` could be matched in css selectively with `[data-state~="tooltip-open"]` or `[data-state~="collapsible-open"]` The `~=` is the important part here. This way a combination of data-state attributes could be done by prefixing and then telling the consumer to match them selectively.

**Reply by benoitgrelard** [reply]
Yeah but that's still prefixing, only the value rather than the key, still not great for DX…

**Reply by BrianHung** [reply]
Wouldn't mind prefixing on key, e.g. `data-tooltip-state` and `data-dropdown-state`. I prefer this over having nested buttons when I want to compose radix components together (tooltip and dropdown below).

```html
<button data-state="closed">
  <button type="button" id="radix-:ri:" aria-haspopup="menu" data-state="closed" class="flex items-center justify-center w-12 h-12 hover:bg-black border-transparent hover:border-gray-200 focus:bg-sky-500 focus:outline-none state-open:bg-sky-500" tabindex="-1" data-orientation="horizontal" data-radix-collection-item="">
  </button>
</button>
```

**Reply by benoitgrelard** [reply]
Nested buttons is invalid anyway so that's out of the question really.

**Comment by jjenzz** [comment]
> Is there any way to make this work?

@PhilGarb As you have realised, you can control all of our components to work around these sorts of issues in the meantime.

```jsx
export default function App() {
  const [open, setOpen] = React.useState<boolean>();

  return (
    <Collapsible.Root open={open} onOpenChange={setOpen}>
      <Tooltip.Root>
        <Tooltip.Trigger
          as={Collapsible.Button}
          className={open ? 'open' : 'closed'} // or use `data-collapsible-state` if that's preferred :)
        >
          Open
        </Tooltip.Trigger>
        <Tooltip.Content>Tooltip</Tooltip.Content>
      </Tooltip.Root>
      <Collapsible.Content>Content</Collapsible.Content>
    </Collapsible.Root>
  );
}
```

_https://codesandbox.io/s/adoring-grass-xtukv?file=/src/App.tsx_

@benoitgrelard This has made me realise though that I had to do `as={Collapsible.Button}` so that the `data-state` is maintained from `Tooltip.Trigger` here. I found that a bit confusing. I initially expected that if I was telling a component to be something else, then the `as` component internal props would take precedence so I tried this at first:

```diff
export default function App() {
  const [open, setOpen] = React.useState<boolean>();

  return (
    <Collapsible.Root open={open} onOpenChange={setOpen}>
      <Tooltip.Root>
-        <Tooltip.Trigger
+        <Collapsible.Button
-          as={Collapsible.Button}
+          as={Tooltip.Trigger}
          className={open ? 'open' : 'closed'}
        >
          Open
-         </Tooltip.Trigger>
+        </Collapsible.Button>
        <Tooltip.Content>Tooltip</Tooltip.Content>
      </Tooltip.Root>
      <Collapsible.Content>Content</Collapsible.Content>
    </Collapsible.Root>
  );
}
```

Something to investigate perhaps?

**Reply by benoitgrelard** [reply]
Hmmm, yeah perhaps. I agree I would expect everything to kinda work the same whichever the order I compose them.

**Reply by jjenzz** [reply]
They can't work the same whichever way they're composed. One will always override the clashing props but I expected the `as` component would be the one that overrides (so `data-state` would be from `Tooltip.Trigger` if `as={Tooltip.Trigger}`) but that wasn't the case 🤔

**Reply by benoitgrelard** [reply]
Yeah that’s what I meant. Obviously here there’s a clash so one would always win. But yes I would expect that to change as well.

**Reply by jjenzz** [reply]
One to add to the list of DX improvements, it really confused me 😅

**Comment by giladv** [comment]
So i also encountered this issue. cannot combine a toggle with a tooltip. 
i might be stating the obvious but why using a custom data-state attr instead of className?
a className can contain multiple classes, AND css selectors are designed to target just a single or multiple of the classes.

am i missing something?

a resulting element could look like:
```
<button class="radix-tooltip-open radix-toggle-on">..</button>
```
it fixes everything, the DX is far better than having to manage the state and it's a far more standard way of doing things IMO.

**Reply by jjenzz** [reply]
Data attributes support multiple values too which @PhilGarb had already suggested:

> So something like: `data-state="tooltip-open collapsibile-open"` could be matched in css selectively with `[data-state~="tooltip-open"]` or `[data-state~="collapsible-open"]` The `~=` is the important part here.

I personally prefer to separate the state into a data attribute as it is easier for us to communicate public vs private API. If everything is a class, it's not clear which classes are public API and safe for you to use, vs private API that could be refactored and stripped from under you. The `data-state` attribute is always public API, everything else you see in the DOM is private.

I'm wondering if the simplest approach would be chaining the `data-state` values and provide a `statePrefix` prop. By default, there would be no prefix so combining two would end up with `data-state="open open"` but that can be rectified with the prefix prop. The simplicity of our current API would remain for the use-cases that aren't combined.

It would be an easy enough change for us too as I think the `statePrefix` can be added to `Primitive`, and `Slot` can be responsible for chaining `data-state` values.

**Reply by giladv** [reply]
hi, thanks for the quick reply!
i don't really mind class vs data-attr honestly as long as there's a solution on the roadmap haha.
is a solution planned for the next version?

thanks!

**Reply by funwithtriangles** [reply]
@jjenzz I think your solution makes a lot of sense. Has an issue been made for this? Can't find one...

**Reply by funwithtriangles** [reply]
Right now I am having the same issue with toggles, but just using `[aria-pressed=true]` as a selector. Seems like a working temporary solution, unless there is some big disadvantage to doing this?

**Reply by jjenzz** [reply]
> Has an issue been made for this? Can't find one...

@funwithtriangles I don't think so. Feel free to raise one and we will prioritise it when we can 🙏

**Reply by lucastobrazil** [reply]
Just sharing my 2c here - as i had the same issue recently (and [discussion in discord](https://discord.com/channels/752614004387610674/803656530259738674/994518619931279400)) - IMO each `data-*` attribute should also contain the radix component name. eg. `data-tooltip-state` and `data-switch-state` for instance.

@benoitgrelard mentioned a few DX trade-offs (my assumption being that a consistent data attribute means you dont need to think about what each component's state selector would be) - which personally I wonder if they're actually a big deal (eg. instead of having the _exact_ same data attributes for all components, you just need to know that the primitive component name is in there too. 

Wouldn't this also make CSS selectors much more communicative too? 

```jsx
const MyComponentThatWantsToStyleOtherComponentsUsingDataAttrs = styled('div', {
    '[data-tooltip-state="open"]': {
        // clearly targeting the tooltip's open state
    }
})
```

Happy to file a specific issue around this topic if need be

**Reply by benoitgrelard** [reply]
Yeah it does make things more explicit upfront for sure.
Thanks for your valuable feedback!

**Reply by MaxMusing** [reply]
+1 for @lucastobrazil's suggestion, this is something our team's been wanting too. We commonly need to compose multiple Radix components which leads to clashing `data-state`s.

**Reply by lucastobrazil** [reply]
@benoitgrelard should I create an issue for this one?

**Reply by benoitgrelard** [reply]
We already have one #602

**Reply by lucastobrazil** [reply]
I'm not sure that actually relates to the topic here?

**Reply by benoitgrelard** [reply]
How so? This is about the issue of `data-state` clashes when composing multiple primitives. Isn't this what is being discussed in this discussion?

**Comment by hipstersmoothie** [comment]
I'm having the similar issue with Popover + Tooltip.

To get the Popover + Tooltip to work I has to use `as="div"` to get the events to bubble correctly. With `asChild` it doesn't work because of the component structure. The direct child is the `Tooltip` and not the `TooltipTrigger` so I think that's why that doesn't work.

```tsx
const MixerButton = () => (
  <Tooltip>
    <TooltipTrigger asChild>
      <IconButton aria-label="Update dimensions">
        <MixerHorizontalIcon />
      </IconButton>
    </TooltipTrigger>
    <TooltipContent sideOffset={5}>Add to library</TooltipContent>
  </Tooltip>
);

const PopoverDemo = () => (
  <Popover>
    <PopoverTrigger as="div">
      <MixerButton />
    </PopoverTrigger>
    <PopoverContent sideOffset={5}>
      <Tooltip>
        <TooltipTrigger>here</TooltipTrigger>
        <TooltipContent sideOffset={5}>Testing</TooltipContent>
      </Tooltip>
    </PopoverContent>
  </Popover>
);
```

If I restructure the markup to the following:

```tax
   <Tooltip>
      <PopoverTrigger asChild>
        <TooltipTrigger asChild>
          <IconButton aria-label="Update dimensions">
            <MixerHorizontalIcon />
          </IconButton>
        </TooltipTrigger>
      </PopoverTrigger>
      <TooltipContent sideOffset={5}>Add to library</TooltipContent>
    </Tooltip>
```

It works (I.e the tooltip and popover both work) but then that means I can't use an API like the example given in the docs

```tax
 <Tooltip content="Tooltip content">
    <button>Tooltip trigger</button>
  </Tooltip>
```

So to get it to work I'd have to change my components API to match Radix's exactly. The `as` prop does what I want, but I'm nervous that it's been marked as deprecated but seem to be the solution here.

**Reply by benoitgrelard** [reply]
Hi @hipstersmoothie,

I don't think your issue is the same as the one discussed here (specifically clash of `data-state`).
In your case I believe your issue is that if you want to close over the `MixerButton` implementation you need to make sure it's public API behaves like a `button`, that means forwarding props and ref to it.

Here's a quick updated sandbox showing it working together nicely:
https://codesandbox.io/s/tooltip-popover-kyddq?file=/src/App.js

**Reply by billylittlefield** [reply]
I ran into this same issue when using a custom wrapper around `Tooltip` and trying to make that the child of a `Dropdown.Trigger`. This answer helped, and I changed my API so my custom tooltip now looks something like this...

```ts
export const CompanyTooltip = React.forwardRef(
    (
        {
            children,
            content,
            ...restOfProps
        }: {
            children: React.ReactNode;
            content: React.ReactNode;
        },
        ref: React.Ref<HTMLButtonElement>
    ) => {
        return (
            <RadixTooltip.Root>
                <RadixTooltip.Trigger asChild ref={ref} {...restOfProps}>
                    {children}
                </RadixTooltip.Trigger>
                <RadixTooltip.Portal>
                    <RadixTooltip.Content>
                        {content}
                    </RadixTooltip.Content>
                </RadixTooltip.Portal>
            </RadixTooltip.Root>
        );
    }
);
```

Now, when using this component inside a `<Dropdown.Trigger asChild />`, the dropdown trigger's props are correctly passed along. 

While this issue (required ref/props-forwarding) is separate from this discussion's OP issue (data-attr clashing), the latter still comes into play here. For instance, in @benoitgrelard 's linked codesandbox when the tooltip is open, the `data-state` shows `"closed"` for the `ButtonWithTooltip` nested in a popover, but `"delayed-open"` for the standalone `ButtonWithTooltip`, because the Popover data-attrs win. This makes sense, and doesn't seem to actually affect the desired behavior, but wanted to call it out. I believe everything's still fine from an a11y perspective, because I see the `aria-describedby` being applied in both cases

Anyway, thanks for the useful discussion in this thread! I've been enjoying learning the Radix toolkit & seeing the active community.

**Comment by tarngerine** [comment]
I ran into this with Toolbar ToggleItems/Buttons, solution was to add a span around the trigger to prevent it creating two nested button OR mangling data-state

```
<Tooltip.Trigger asChild>
  <span>
    {children}
  </span>
</Tooltip.Trigger>
```

**Reply by jackbuscombe** [reply]
Thank u based god 🙏

**Reply by asfires** [reply]
Thanks so much for sharing this! I had this exact issue but this totally works and fixed it. In my case I had to add the asChild prop to my TooltipTrigger component and wrap my {item.key} in a span.

```jsx
<TooltipProvider delayDuration={550}>
  <Tooltip>
    <FormLabel className="font-normal">
      <TooltipTrigger asChild>
        <span>{item.key}</span>
      </TooltipTrigger>
      <TooltipContent className="max-w-md">
        <p>{item.description}</p>
      </TooltipContent>
    </FormLabel>
  </Tooltip>
</TooltipProvider>
```

**Reply by Natan4Securitas** [reply]
Thanks, this worked for me too

**Reply by emillaine** [reply]
This breaks accessibility, since the `aria-describedby` now gets added to the `span` instead of the focusable trigger element. So the tooltip doesn't get announced by screen readers.

**Reply by cgatian** [reply]
I'm putting this up for some 👀.
Taking @tarngerine suggestion and the feedback from @emlai, I made a component that only appends the `span` element when `data-state` is provided. Other props like `aria-describedby` are passed to the child.

Seems to work, but I don't know if I'm missing an edge case.

```tsx
import { Slot } from '@radix-ui/react-slot';
import { ComponentProps, ElementType, forwardRef } from 'react';

const DataStatePropInterceptor = forwardRef<HTMLElement, ComponentProps<ElementType>>(
  (props, ref) => {
    const { 'data-state': dataState, children, ...rest } = props;
    if (dataState) {
      return (
        <span data-state={dataState}>
          <Slot {...rest} ref={ref}>
            {children}
          </Slot>
        </span>
      );
    }
    return (
      <Slot {...rest} ref={ref}>
        {children}
      </Slot>
    );
  },
);
```

Wrap the `TooltipPrimitive.Trigger` with it

```tsx
<TooltipPrimitive.Trigger asChild>
   <DataStatePropInterceptor>{children}</DataStatePropInterceptor>
</TooltipPrimitive.Trigger>
```

**Reply by tomgreeEn** [reply]
@cgatian this is the only thing that works for me, thanks

**Comment by WilliamIPark** [comment]
Has any progress been made on resolving the `data-state` clash?

**Comment by dcastil** [comment]
Just chiming in with a solution I settled on when using multiple Radix components on the same trigger, like a tooltip and a dropdown menu on the same button.

My use case is that I have custom design system components like `MyDropdownMenu` and `MyTooltip` which both use Radix under the hood. That means that both will attach the trigger props to their children like the `ref` for the trigger, resulting in `MyDropdownMenu` passing those props onto `MyTooltip` instead of the intended `Button`.

```tsx
// Problem
function MyButtonWithTooltipAndDropdown() {
    return (
        // ↓ MyDropdownMenu passes Radix props onto MyTooltip instead of intended Button
        <MyDropdownMenu>
            <MyTooltip>
                <Button>Click me!</Button>
            </MyTooltip>
        </MyDropdownMenu>
    )
}
```

The problem here is that `MyTooltip` already uses the `ref` and its rest properties for the tooltip content and therefore can't pass those down to the `Button`. The `MyDropdownMenu` component needs to "skip" the `MyTooltip` component and attach the props on the child of `MyTooltip` instead.

For this I created a `ForwardPropsToChild` component which does exactly that: It passes its own props to the child of its child.

```tsx
import { Slot } from '@radix-ui/react-slot'
import { cloneElement, forwardRef } from 'react'

interface ForwardPropsToChildProps {
    children: React.ReactElement<{ children: React.ReactNode }>
}

export const ForwardPropsToChild = forwardRef<HTMLElement, ForwardPropsToChildProps>(
    function ForwardPropsToChild({ children, ...props }, ref) {
        return cloneElement(
            children,
            undefined,
            // Slot used so that callbacks, className, style, etc. are composed together instead of replaced.
            <Slot ref={ref} {...props}>
                {children.props.children}
            </Slot>,
        )
    },
)
```

This then can be used like the following code snippet to forward props to the intended component.

```tsx
// Solution
function MyButtonWithTooltipAndDropdown() {
    return (
        // ↓ MyDropdownMenu passes Radix props onto ForwardPropsToChild
        <MyDropdownMenu>
            {/* ↓ ForwardPropsToChild forwards Radix props to Button */}
            <ForwardPropsToChild>
                <MyTooltip>
                    <Button>Click me!</Button>
                </MyTooltip>
            </ForwardPropsToChild>
        </MyDropdownMenu>
    )
}
```

I think it would be nicer to be able to get the props with a `children` function from Radix to pass the children props to another element in a more straight forward and explicit way. But so far the `ForwardPropsToChild` component is the least ugly solution I could come up with to mitigate the current issue.

Side note: This doesn't solve the issue with clashing states like `data-state`, but works well for cases where inserting another node in between elements makes layouting difficult.

**Reply by castrosuellenx** [reply]
Thanks for this

**Comment by nklhtv** [comment]
To avoid collisions, `data-state` attributes can be scoped by id passed as a prop, e.g. by using React's useId() hook.
This behavior could be implemented as opt-in, so that its not a breaking change. If the primitive component receives no `id` prop then the `state-open` data attr will be used, otherwise `state-${id}-open` data attr will be used.

**Comment by Raepheles** [comment]
Is there still no progress on this? I wanted to have tooltip for toggle component and I was faced with the same problem. Toggle's "data-state" property doesn't update therefor classes based on data-state don't apply.

```tsx
<TooltipProvider>
  <Tooltip>
    <TooltipTrigger asChild>
       <Toggle pressed={pressed} onPressedChange={setPressed}>
         <FontItalicIcon />
       </Toggle>
    </TooltipTrigger>
    <TooltipContent>
      <p>Tooltip</p>
    </TooltipContent>
  </Tooltip>
</TooltipProvider>
```

**Comment by raythurnvoid** [comment]
This works
```tsx
    <SybPopover open={isOpen} onOpenChange={handleOpenChange}>
      <SybTooltip>
        <RadixTooltip.Trigger asChild>
          <RadixPopover.Trigger asChild>
            <button aria-pressed={isActive}>test</button>
          </RadixPopover.Trigger>
        </RadixTooltip.Trigger>
        <SybTooltipPopover sideOffset="md">
          <SybTooltipText>Link</SybTooltipText>
        </SybTooltipPopover>
      </SybTooltip>
 {/* rest... */}
 ```
 
 This does not
 ```tsx
    <SybPopover open={isOpen} onOpenChange={handleOpenChange}>
      <SybTooltip>
        <SybTooltipTrigger>  {/* my custom wrapper */}
          <SybPopoverTrigger>  {/* my custom wrapper */}
            <button size="sm" variant="tertiary-gray" aria-label="Link" aria-pressed={isActive}>
              test
            </button>
          </SybPopoverTrigger>
        </SybTooltipTrigger>
        <SybTooltipPopover sideOffset="md">
          <SybTooltipText>Link</SybTooltipText>
        </SybTooltipPopover>
      </SybTooltip>
      
 {/* rest... */}
 
 // SybTooltipTrigger.tsx
 export function SybTooltipTrigger(props) {
    const { children, ...rest } = props;

   return (
    <RadixTooltip.Trigger asChild {...rest}>
      {children}
    </RadixTooltip.Trigger>
  );
}

// SybPopoverTrigger.tsx
export function SybPopoverTrigger(props: SybPopoverTrigger.Props) {
  const { children, ...rest } = props;

  return (
    <RadixPopover.Trigger asChild {...rest}>
      {children}
    </RadixPopover.Trigger>
  );
}
```

**Comment by giladv** [comment]
4 years after - Is there any news on this?

**Reply by robertoms99** [reply]
if you tree is` TooltipTrigger(asChild) > Toggle` ,  the resulting html element will have TooltipTrigger's data-state values, if you invert the parenting, will be the opposited. I think we need to play with this for now. But, if you prefer you can use an data-state's interceptor
